File: src/og_render.py
#!/usr/bin/env python3
"""
og_render.py - convertit les images de partage (Open Graph) de SVG en PNG.

Appele par build_site.js juste apres la generation du site, via :
    python3 og_render.py

build_site.js ecrit un .svg par page dans site/og/<nom>.svg (fonction
queueOg()) mais le HTML genere pointe toujours vers <nom>.png dans les
balises <meta property="og:image">, jamais vers le .svg (les plateformes
sociales - Facebook, LinkedIn, Twitter/X - n'acceptent pas le SVG comme
image de partage). Ce script comble ce trou : il lit chaque .svg present
dans site/og/, le convertit en .png de meme nom a cote, et laisse le .svg
en place (inoffensif, non reference par le HTML).

Dependance : cairosvg (et sa lib systeme libcairo2).
    pip install cairosvg --break-system-packages
Si Debian/Ubuntu sans libcairo2 : sudo apt install libcairo2

Ce script ne fait rien de destructif : si un .png existe deja et est plus
recent que son .svg source, il est saute (reconstruction plus rapide en
developpement). Utiliser --force pour tout reconverfir.
"""
import os
import sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
OG_DIR = os.path.join(SCRIPT_DIR, "site", "og")


def main():
    force = "--force" in sys.argv

    if not os.path.isdir(OG_DIR):
        print(f"og_render.py : dossier introuvable ({OG_DIR}), rien a faire.")
        return 0

    try:
        import cairosvg
    except ImportError:
        print("og_render.py : le module 'cairosvg' est absent.")
        print("  Installer avec : pip install cairosvg --break-system-packages")
        print("  (necessite aussi la lib systeme libcairo2 : apt install libcairo2)")
        return 1

    svg_files = sorted(f for f in os.listdir(OG_DIR) if f.endswith(".svg"))
    if not svg_files:
        print("og_render.py : aucun .svg trouve dans site/og/.")
        return 0

    converted = 0
    skipped = 0
    failed = 0

    for fname in svg_files:
        svg_path = os.path.join(OG_DIR, fname)
        png_path = os.path.join(OG_DIR, fname[:-4] + ".png")

        if not force and os.path.isfile(png_path) and \
                os.path.getmtime(png_path) >= os.path.getmtime(svg_path):
            skipped += 1
            continue

        try:
            cairosvg.svg2png(url=svg_path, write_to=png_path,
                              output_width=1200, output_height=630)
            converted += 1
        except Exception as e:
            print(f"  echec sur {fname} : {e}")
            failed += 1

    print(f"og_render.py : {converted} converties, {skipped} deja a jour, "
          f"{failed} echecs (sur {len(svg_files)} .svg).")
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(main())

File: src/og_render.js
#!/usr/bin/env node
/* og_render.js - convertit les images de partage (Open Graph) de SVG en PNG.

   Remplace og_render.py, appele par build_site.js juste apres la generation
   du site. Meme role, memes fichiers, meme ligne de compte-rendu.

   Pourquoi ne plus passer par Python. Le script Python demandait cairosvg,
   qui demande lui-meme la bibliotheque systeme Cairo. Sur Linux c'est une
   ligne d'apt ; sur Windows c'est une installation de Python, puis de
   cairosvg, puis des DLL de GTK, et build_site.js appelait de surcroit
   "python3", un nom qui n'existe pas sur une installation Windows standard.
   sharp arrive avec son moteur de rendu compile pour la plateforme :
   npm install sharp, et c'est fini.

   Fidelite verifiee avant bascule, en comparant pixel a pixel les rendus des
   deux outils sur les memes SVG : ecart moyen de 1 sur 255, 1,4 % des octets
   au-dela de 8, uniquement sur les bords de lettres. Les aplats, les couleurs
   et la geometrie sont identiques.

   build_site.js ecrit un .svg par page dans site/og/<nom>.svg, mais le HTML
   pointe vers <nom>.png : les plateformes sociales n'acceptent pas le SVG en
   image de partage. Ce script comble ce trou et laisse le .svg en place.

   Rien de destructif : un .png plus recent que son .svg est saute.
   --force reconvertit tout.
*/
const fs = require("fs");
const path = require("path");

const OG_DIR = path.join(__dirname, "site", "og");
const force = process.argv.includes("--force");

let sharp;
try {
  sharp = require("sharp");
} catch (e) {
  console.log("og_render.js : le module 'sharp' est absent.");
  console.log("  Installer avec : npm install sharp --no-save");
  process.exit(1);
}

if (!fs.existsSync(OG_DIR)) {
  console.log("og_render.js : dossier introuvable (" + OG_DIR + "), rien a faire.");
  process.exit(0);
}

const svgs = fs.readdirSync(OG_DIR).filter(f => f.endsWith(".svg")).sort();
if (!svgs.length) {
  console.log("og_render.js : aucun .svg trouve dans site/og/.");
  process.exit(0);
}

(async () => {
  let converties = 0, sautees = 0, echecs = 0;

  for (const nom of svgs) {
    const svg = path.join(OG_DIR, nom);
    const png = path.join(OG_DIR, nom.slice(0, -4) + ".png");

    if (!force && fs.existsSync(png) &&
        fs.statSync(png).mtimeMs >= fs.statSync(svg).mtimeMs) { sautees++; continue; }

    try {
      /* La taille est imposee, pas deduite du SVG : les plateformes sociales
         attendent 1200x630 et redimensionnent mal ce qui s'en ecarte. */
      await sharp(svg, { density: 96 })
        .resize(1200, 630)
        .png()
        .toFile(png);
      converties++;
    } catch (err) {
      console.log("  echec sur " + nom + " : " + err.message);
      echecs++;
    }
  }

  console.log("og_render.js : " + converties + " converties, " + sautees +
    " deja a jour, " + echecs + " echecs (sur " + svgs.length + " .svg).");
  process.exit(echecs ? 1 : 0);
})();
